Add Spotify to the desktop class quirks list

brentfulgham opened the pull request:
#### c710e9dd3240943ddeca3f902c0b7aefbbf15caa
<pre>
Add Spotify to the desktop class quirks list
<a href="https://bugs.webkit.org/show_bug.cgi?id=268928">https://bugs.webkit.org/show_bug.cgi?id=268928</a>
&lt;<a href="https://rdar.apple.com/121414876">rdar://121414876</a>&gt;

Reviewed by Wenson Hsieh and Chris Dumez.

It is nearly impossible to use the volume control or playback scrubber
in Spotify in Safari. They don&apos;t expect a desktop with touch support.

Spotify removed the scrubber features on iOS back in 2016 (<a href="https://community.spotify.com/t5/iOS-iPhone-iPad/Audio-scrubber-is-gone-in-latest-update-Why/td-p/1311946">https://community.spotify.com/t5/iOS-iPhone-iPad/Audio-scrubber-is-gone-in-latest-update-Why/td-p/1311946</a>

This patch presents Safari to Spotify as a mobile User Agent so that we
get the best &quot;touch&quot; version of the website.

* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::desktopClassBrowsingRecommendedForRequest):

Canonical link: <a href="https://commits.webkit.org/274268@main">https://commits.webkit.org/274268@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c57232ed5a26b8b9f8cc02c1e333da57037961fb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38444 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17388 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40783 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40995 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/34161 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20165 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14732 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32367 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39017 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14667 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33484 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12746 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/12728 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42271 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/34951 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38565 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13324 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11023 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36767 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14904 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/13792 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5013 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14376 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

karlcow commented on Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
```diff
@@ -1423,6 +1423,10 @@ static RecommendDesktopClassBrowsingForRequest desktopClassBrowsingRecommendedFo
         return RecommendDesktopClassBrowsingForRequest::No;
 
+    // FIXME: Remove this quirk when <rdar://122481999> is complete
+    if (equalLettersIgnoringASCIICase(host, "spotify.com"_s) || host.endsWithIgnoringASCIICase(".spotify.com"_s))
```
@brentfulgham 

```suggestion
    if (equalLettersIgnoringASCIICase(host, "spotify.com"_s) || host.endsWithIgnoringASCIICase(".spotify.com"_s) || equalLettersIgnoringASCIICase(host, "open.spotifycdn.com"_s))
```

All the JS are on `open.spotifycdn.com`.

whsieh replied:
Is `open.spotifycdn.com` the host of the top URL?

If not, then I think this part of the quirk would only apply when opening the JavaScript file itself as a main resource (i.e. viewing the JS source as plain text).

brentfulgham committed:
Add Spotify to the desktop class quirks list
https://bugs.webkit.org/show_bug.cgi?id=268928
<rdar://121414876>

Reviewed by Wenson Hsieh and Chris Dumez.

It is nearly impossible to use the volume control or playback scrubber
in Spotify in Safari. They don't expect a desktop with touch support.

Spotify removed the scrubber features on iOS back in 2016 (https://community.spotify.com/t5/iOS-iPhone-iPad/Audio-scrubber-is-gone-in-latest-update-Why/td-p/1311946

This patch presents Safari to Spotify as a mobile User Agent so that we
get the best "touch" version of the website.

* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::desktopClassBrowsingRecommendedForRequest):

Canonical link: https://commits.webkit.org/274268@main
